integration/setup/setup.go

```diff
@@ -3,33 +3,41 @@
 package setup
 
 import (
-	"log"
+	"context"
+	"fmt"
 	"os"
 	"testing"
 
-	"github.com/giantswarm/e2e-harness/pkg/framework"
+	"github.com/giantswarm/helmclient/integration/env"
+	"github.com/giantswarm/microerror"
 )
 
-func WrapTestMain(h *framework.Host, m *testing.M) {
-	var v int
-	var err error
+func Setup(m *testing.M, config Config) {
+	ctx := context.Background()
 
-	if err = h.CreateNamespace("giantswarm"); err != nil {
-		log.Printf("%#v\n", err)
-		v = 1
+	exitCode, err := setup(ctx, m, config)
+	if err != nil {
+		config.Logger.LogCtx(ctx, "level", "error", "message", "", "stack", fmt.Sprintf("%#v", err))
+		os.Exit(1)
 	}
 
-	if v == 0 {
-		m.Run()
-	}
+	os.Exit(exitCode)
+}
 
-	if os.Getenv("KEEP_RESOURCES") != "true" {
-		// only do full teardown when not on CI
-		if os.Getenv("CIRCLECI") != "true" {
-			// TODO there should be error handling for the framework teardown.
-			h.Teardown()
+func setup(ctx context.Context, m *testing.M, config Config) (int, error) {
+	var err error
+	teardown := !env.CircleCI() && !env.KeepResources()
+
+	{
+		// TODO this should be named EnsureNamespaceCreated
+		err = config.K8sSetup.EnsureNamespace(ctx, tillerNamespace)
+		if err != nil {
+			return 1, microerror.Mask(err)
+		}
+		if teardown {
+			// TODO defer config.K8sSetup.EnsureNamespaceDeleted(ctx, tillerNamespace)
```

Comment:
I was talking about that already. `defer` instead of `v == 1`. Deferred functions are pushed to stack so we can be sure deletion will happen in desired order.